server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java

```diff
@@ -505,38 +520,40 @@ public String toString() {
             }
         }
 
-        int numberOfThreads = randomIntBetween(2, 8);
-        int taskSubmissionsPerThread = randomIntBetween(1, 64);
-        int numberOfExecutors = Math.max(1, numberOfThreads / 4);
-        final Semaphore semaphore = new Semaphore(numberOfExecutors);
+        final int numberOfThreads = randomIntBetween(2, 8);
+        final int taskSubmissionsPerThread = randomIntBetween(1, 64);
+        final int numberOfExecutors = Math.max(1, numberOfThreads / 4);
+        final Semaphore semaphore = new Semaphore(1);
 
         class TaskExecutor implements ClusterStateTaskExecutor<Task> {
-            private final List<Set<Task>> taskGroups;
-            private AtomicInteger counter = new AtomicInteger();
-            private AtomicInteger batches = new AtomicInteger();
-            private AtomicInteger published = new AtomicInteger();
 
-            TaskExecutor(List<Set<Task>> taskGroups) {
-                this.taskGroups = taskGroups;
-            }
```

Comment:
Moved this to `assignments`. Now executor will verify it `only` executes own tasks

Reply:
I think we don't actually verify this yet, we're now only checking that every group _of our own tasks_ is completely executed. Previously we were checking that every group _of any executor's tasks_ is completely executed so technically this is now a weaker test.

I think this would be a good property to verify, but to do this we also need to say that every task we're executing belongs to one of our own groups.

Reply:
I just added `assertThat("All tasks should belong to this executor", totalCount, equalTo(tasks.size()));` to verify that every single task from the list belongs to the current executor.

There is also the a check in the very end of the test that `assertEquals(executor.assigned.get(), executor.executed.get())` that verifies that all assigned tasks are executed. We also have a check that no task is executed twice.

I believe this should cover everything, is it?